File: packages/python/stagehand_social/core/base_platform.py
"""Abstract BasePlatform — mirror of packages/typescript/src/core/BasePlatform.ts."""

from __future__ import annotations

import os
import time
from abc import ABC, abstractmethod
from typing import Optional

from stagehand import Stagehand, StagehandConfig

from stagehand_social.core.errors import AuthenticationError, ConfigError
from stagehand_social.core.session import SessionManager
from stagehand_social.core.types import (
    CommentInput,
    CommentResult,
    Credentials,
    FollowInput,
    FollowResult,
    LikeInput,
    LikeResult,
    LLMConfig,
    MfaCallback,
    PlatformName,
    PlatformTarget,
    PostInput,
    PostResult,
)


class BasePlatform(ABC):
    """Abstract base class for specialized platform adapters."""

    def __init__(
        self,
        platform: PlatformName,
        account: str,
        session_manager: SessionManager,
        llm: LLMConfig,
        credentials: Optional[Credentials] = None,
        on_mfa_challenge: Optional[MfaCallback] = None,
        headless: bool = True,
        log_level: str = "info",
    ):
        self.platform: PlatformName = platform
        self.account = account
        self.session_manager = session_manager
        self.credentials = credentials
        self.on_mfa_challenge = on_mfa_challenge
        self.llm_config = llm
        self.headless = headless
        self.log_level = log_level
        self._stagehand: Optional[Stagehand] = None

    # ------------------------------------------------------------------------
    # Subclass contract
    # ------------------------------------------------------------------------

    @abstractmethod
    def login_url(self) -> str: ...

    @abstractmethod
    async def perform_login(self, creds: Credentials) -> None: ...

    @abstractmethod
    async def is_logged_in(self) -> bool: ...

    @abstractmethod
    async def post(self, input: PostInput) -> PostResult: ...

    @abstractmethod
    async def comment(self, input: CommentInput) -> CommentResult: ...

    @abstractmethod
    async def like(self, input: LikeInput) -> LikeResult: ...

    @abstractmethod
    async def unlike(self, input: LikeInput) -> LikeResult: ...

    @abstractmethod
    async def follow(self, input: FollowInput) -> FollowResult: ...

    @abstractmethod
    async def unfollow(self, input: FollowInput) -> FollowResult: ...

    # ------------------------------------------------------------------------
    # Shared concrete methods
    # ------------------------------------------------------------------------

    async def get_stagehand(self) -> Stagehand:
        if self._stagehand is not None:
            return self._stagehand
        target = PlatformTarget(platform=self.platform)
        context_id = self.session_manager.get_or_create_context(target, self.account)
        session_params = self.session_manager.build_session_params(context_id)

        config = StagehandConfig(
            env="BROWSERBASE",
            project_id=session_params["project_id"],
            browserbase_session_create_params=session_params,
            model_name=self.llm_config.model,
            model_client_options={"apiKey": self.llm_config.api_key},
            verbose=2 if self.log_level == "debug" else (1 if self.log_level == "info" else 0),
            headless=self.headless,
        )
        self._stagehand = Stagehand(config)
        await self._stagehand.init()
        return self._stagehand

    async def login(self, creds: Optional[Credentials] = None) -> None:
        resolved = creds or self.credentials or self._load_creds_from_env()
        if resolved is None:
            raise ConfigError(
                f"No credentials for {self.platform}. Set "
                f"STAGEHAND_SOCIAL_{self.platform.upper()}_USERNAME / _PASSWORD "
                "or pass them to login()."
            )
        self.credentials = resolved
        await self.get_stagehand()

        try:
            already = await self.is_logged_in()
        except Exception:
            already = False
        if already:
            self.session_manager.mark_verified(PlatformTarget(platform=self.platform), self.account)
            return

        sh = await self.get_stagehand()
        await sh.page.goto(self.login_url())
        await self.perform_login(resolved)

        if not await self.is_logged_in():
            raise AuthenticationError(
                f"{self.platform} login appeared to succeed but is_logged_in() returned False."
            )
        self.session_manager.mark_verified(PlatformTarget(platform=self.platform), self.account)

    async def close(self) -> None:
        if self._stagehand is None:
            return
        try:
            await self._stagehand.close()
        finally:
            self._stagehand = None

    async def logout(self) -> None:
        sh = await self.get_stagehand()
        try:
            await sh.page.act("Open the user menu and click 'Log out'")
        finally:
            self.session_manager.forget(PlatformTarget(platform=self.platform), self.account)
            await self.close()

    # ------------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------------

    def _load_creds_from_env(self) -> Optional[Credentials]:
        prefix = f"STAGEHAND_SOCIAL_{self.platform.upper()}"
        username = os.environ.get(f"{prefix}_USERNAME")
        password = os.environ.get(f"{prefix}_PASSWORD")
        totp_secret = os.environ.get(f"{prefix}_TOTP_SECRET")
        if not username or not password:
            return None
        return Credentials(username=username, password=password, totp_secret=totp_secret)

    def _now(self) -> int:
        return int(time.time() * 1000)

File: packages/typescript/src/core/BasePlatform.ts
/**
 * Abstract base class for all specialized platform adapters.
 *
 * Each platform (instagram, tiktok, x, etc.) extends this and implements
 * the abstract methods. Stagehand is owned by this class — subclasses just
 * issue `act()` / `extract()` calls via `this.stagehand`.
 *
 * Lifecycle:
 *   1. SocialSDK.create() → instantiates an adapter with a SessionManager.
 *   2. adapter.login() → opens a Stagehand session against the platform's
 *      Browserbase context, drives the login flow, marks the session verified.
 *   3. adapter.post() / .comment() / etc. → reuses the open session.
 *   4. adapter.close() → tears down the Stagehand session (context persists).
 */

import { Stagehand } from "@browserbasehq/stagehand";
import type {
  Credentials,
  CommentInput,
  CommentResult,
  FollowInput,
  FollowResult,
  LikeInput,
  LikeResult,
  MfaCallback,
  PlatformName,
  PostInput,
  PostResult,
  TargetKey,
} from "./types.js";
import { AuthenticationError, ConfigError } from "./errors.js";
import { SessionManager } from "./session.js";

export interface BasePlatformOpts {
  platform: PlatformName;
  account: string;
  sessionManager: SessionManager;
  credentials?: Credentials;
  onMfaChallenge?: MfaCallback;
  llm: { provider: "anthropic" | "openai"; apiKey: string; model?: string };
  headless?: boolean;
  logLevel?: "debug" | "info" | "warn" | "error";
}

export abstract class BasePlatform {
  /** Sub-domain of social media this adapter handles. */
  public readonly platform: PlatformName;

  /** Account label — used to disambiguate multi-account setups. */
  public readonly account: string;

  /** Built lazily on first action. */
  protected stagehand?: Stagehand;

  /** Resolved credentials (env vars or constructor override). */
  protected credentials?: Credentials;

  protected readonly sessionManager: SessionManager;
  protected readonly onMfaChallenge?: MfaCallback;
  protected readonly llmConfig: BasePlatformOpts["llm"];
  protected readonly headless: boolean;
  protected readonly logLevel: NonNullable<BasePlatformOpts["logLevel"]>;

  constructor(opts: BasePlatformOpts) {
    this.platform = opts.platform;
    this.account = opts.account;
    this.sessionManager = opts.sessionManager;
    this.credentials = opts.credentials;
    this.onMfaChallenge = opts.onMfaChallenge;
    this.llmConfig = opts.llm;
    this.headless = opts.headless ?? true;
    this.logLevel = opts.logLevel ?? "info";
  }

  // -------------------------------------------------------------------------
  // Subclass contract — must implement
  // -------------------------------------------------------------------------

  /** URL of the platform's main login page. */
  protected abstract loginUrl(): string;

  /** Drive the login flow once Stagehand is on the login page. */
  protected abstract performLogin(creds: Credentials): Promise<void>;

  /** Detect whether the current browser is logged in (heuristic per platform). */
  abstract isLoggedIn(): Promise<boolean>;

  /** Publish a new post. */
  abstract post(input: PostInput): Promise<PostResult>;

  /** Add a comment to an existing post. */
  abstract comment(input: CommentInput): Promise<CommentResult>;

  /** Like a post. */
  abstract like(input: LikeInput): Promise<LikeResult>;

  /** Unlike a post. */
  abstract unlike(input: LikeInput): Promise<LikeResult>;

  /** Follow an account. */
  abstract follow(input: FollowInput): Promise<FollowResult>;

  /** Unfollow an account. */
  abstract unfollow(input: FollowInput): Promise<FollowResult>;

  // -------------------------------------------------------------------------
  // Shared concrete methods
  // -------------------------------------------------------------------------

  /** Get the running Stagehand instance, creating it on first call. */
  protected async getStagehand(): Promise<Stagehand> {
    if (this.stagehand) return this.stagehand;

    const target: TargetKey = { kind: "platform", platform: this.platform };
    const contextId = await this.sessionManager.getOrCreateContext(
      target,
      this.account,
    );
    const sessionParams = this.sessionManager.buildSessionParams(contextId);

    this.stagehand = new Stagehand({
      env: "BROWSERBASE",
      apiKey: undefined, // SDK reads BROWSERBASE_API_KEY from env automatically when env is BROWSERBASE
      projectId: sessionParams.projectId,
      browserbaseSessionCreateParams: sessionParams,
      modelName: this.llmConfig.model,
      modelClientOptions: { apiKey: this.llmConfig.apiKey },
      verbose: this.logLevel === "debug" ? 2 : this.logLevel === "info" ? 1 : 0,
      headless: this.headless,
    });

    await this.stagehand.init();
    return this.stagehand;
  }

  /**
   * Top-level login. Resolves credentials, ensures Stagehand is running,
   * navigates to login URL, calls subclass `performLogin`, verifies, persists.
   */
  async login(creds?: Credentials): Promise<void> {
    const resolved = creds ?? this.credentials ?? this.loadCredsFromEnv();
    if (!resolved) {
      throw new ConfigError(
        `No credentials for ${this.platform}. Set ` +
          `STAGEHAND_SOCIAL_${this.platform.toUpperCase()}_USERNAME / _PASSWORD ` +
          `or pass them to login().`,
      );
    }
    this.credentials = resolved;

    const sh = await this.getStagehand();

    // Already logged in (cookie persisted from a prior run)?
    if (await this.isLoggedIn().catch(() => false)) {
      await this.sessionManager.markVerified(
        { kind: "platform", platform: this.platform },
        this.account,
      );
      return;
    }

    await sh.page.goto(this.loginUrl());
    await this.performLogin(resolved);

    if (!(await this.isLoggedIn())) {
      throw new AuthenticationError(
        `${this.platform} login appeared to succeed but isLoggedIn() returned false.`,
      );
    }

    await this.sessionManager.markVerified(
      { kind: "platform", platform: this.platform },
      this.account,
    );
  }

  /** Tear down the Stagehand session. Browserbase context persists for next run. */
  async close(): Promise<void> {
    if (!this.stagehand) return;
    try {
      await this.stagehand.close();
    } finally {
      this.stagehand = undefined;
    }
  }

  /** Default logout — most subclasses don't need to override. */
  async logout(): Promise<void> {
    const sh = await this.getStagehand();
    try {
      await sh.page.act("Open the user menu and click 'Log out'");
    } finally {
      await this.sessionManager.forget(
        { kind: "platform", platform: this.platform },
        this.account,
      );
      await this.close();
    }
  }

  // -------------------------------------------------------------------------
  // Helpers
  // -------------------------------------------------------------------------

  protected loadCredsFromEnv(): Credentials | undefined {
    const prefix = `STAGEHAND_SOCIAL_${this.platform.toUpperCase()}`;
    const username = process.env[`${prefix}_USERNAME`];
    const password = process.env[`${prefix}_PASSWORD`];
    const totpSecret = process.env[`${prefix}_TOTP_SECRET`];
    if (!username || !password) return undefined;
    return { username, password, totpSecret };
  }

  /** Convenience: timestamp for ActionResult.at. */
  protected now(): number {
    return Date.now();
  }
}
